src/lib/support/ThreadOperationalDataset.cpp

```diff
@@ -287,240 +252,233 @@ CHIP_ERROR OperationalDataset::GetMasterKey(uint8_t (&aMasterKey)[kSizeMasterKey
+    tlv->Get32(aDelayMillis);
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR OperationalDataset::SetPanId(uint16_t aPanId)
-{
-    ThreadTLV * tlv = MakeRoom(ThreadTLV::kPanId, sizeof(*tlv) + sizeof(aPanId));
-    VerifyOrReturnError(tlv != nullptr, CHIP_ERROR_NO_MEMORY);
-
-    tlv->Set16(aPanId);
-
-    mLength = static_cast<uint8_t>(mLength + tlv->GetSize());
+/// OperationalDataset
 
+CHIP_ERROR OperationalDataset::Init(ByteSpan aData)
+{
+    VerifyOrReturnError(IsValid(aData), CHIP_ERROR_INVALID_ARGUMENT);
+    // Use memmove because aData could be a sub-span of AsByteSpan()
+    memmove(mBuffer, aData.data(), aData.size());
+    mData = ByteSpan(mBuffer, aData.size());
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR OperationalDataset::GetPSKc(uint8_t (&aPSKc)[kSizePSKc]) const
+void OperationalDataset::CopyDataIfNecessary()
 {
-    const ThreadTLV * tlv = Locate(ThreadTLV::kPSKc);
-    VerifyOrReturnError(tlv != nullptr, CHIP_ERROR_TLV_TAG_NOT_FOUND);
-    VerifyOrReturnError(tlv->GetLength() == sizeof(aPSKc), CHIP_ERROR_INVALID_TLV_ELEMENT);
-    memcpy(aPSKc, tlv->GetValue(), sizeof(aPSKc));
-    return CHIP_NO_ERROR;
+    // It's possible that mData points into an external buffer if someone has
+    // called OperationalDatasetView::Init() instead of our copying version.
+    if (mData.data() != mBuffer)
+    {
+        CopyData();
+    }
 }
 
-CHIP_ERROR OperationalDataset::SetPSKc(const uint8_t (&aPSKc)[kSizePSKc])
+void OperationalDataset::CopyData()
 {
-    ThreadTLV * tlv = MakeRoom(ThreadTLV::kPSKc, sizeof(*tlv) + sizeof(aPSKc));
-    VerifyOrReturnError(tlv != nullptr, CHIP_ERROR_NO_MEMORY);
+    memmove(mBuffer, mData.data(), mData.size());
+    mData = ByteSpan(mBuffer, mData.size());
+}
 
-    tlv->SetValue(aPSKc, sizeof(aPSKc));
+void OperationalDataset::Remove(ThreadTLV * tlv)
+{
+    size_t size      = tlv->GetSize();
+    ThreadTLV * next = tlv->GetNext();
+    memmove(tlv, next, static_cast<size_t>(mData.end() - reinterpret_cast<uint8_t *>(next)));
+    mData = ByteSpan(mData.data(), mData.size() - size);
+}
+
+void OperationalDataset::Remove(uint8_t aType)
+{
+    CopyDataIfNecessary();
+    ThreadTLV * tlv = const_cast<ThreadTLV *>(Locate(aType));
+    if (tlv != nullptr)
+    {
+        Remove(tlv);
+    }
+}
 
-    mLength = static_cast<uint8_t>(mLength + tlv->GetSize());
+// Inserts a TLV of the specified type and length into the dataset, replacing an existing TLV
+// of the same type if one exists. Returns nullptr if there is not enough space.
+ThreadTLV * OperationalDataset::InsertOrReplace(uint8_t aType, size_t aValueSize)
+{
+    CopyDataIfNecessary();
+    ThreadTLV * tlv = const_cast<ThreadTLV *>(Locate(aType));
+    if (tlv != nullptr)
+    {
+        size_t tlvLength = tlv->GetLength();
+        VerifyOrReturnValue(aValueSize != tlvLength, tlv); // re-use in place if same size
+        VerifyOrReturnValue(aValueSize < tlvLength || mData.size() + aValueSize - tlvLength <= sizeof(mBuffer), nullptr);
```

Comment:
Depending on how much we trust our inputs, this might be a more overflow-safe check:
```suggestion
        VerifyOrReturnValue(aValueSize < tlvLength || aValueSize <= sizeof(mBuffer) - (mData.size() - tlvLength), nullptr);
```

Reply:
This is an internal helper methods and the caller is expected to have already validated aValueSize makes sense for the specific value being set, or checked against the max size (kMaxDatasetElementLength) if there is no other constraint. I've added an `assert` and a comment, so I think we can leave these expressions in the maybe more readable arrangement.